compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,338 @@
+    AffineExpr baseOffsetExpr =
+        subgroupIdExpr * totalCopySizeExpr + iterationExpr * strideExpr;
+    return affine::makeComposedAffineApply(rewriter, loc, baseOffsetExpr,
+                                           {sgIdVal, indVar})
+        .getResult();
+  };
+
+  // Build for loop skeleton:
+  scf::ForOp forOp = rewriter.create<scf::ForOp>(
+      loc, /*lb=*/rewriter.create<arith::ConstantIndexOp>(loc, 0),
+      /*ub=*/rewriter.create<arith::ConstantIndexOp>(loc, numCopiesPerThread),
+      /*steps=*/rewriter.create<arith::ConstantIndexOp>(loc, 1));
+
+  auto delinearizeIndex = [&](Value index, ArrayRef<int64_t> shape) {
+    return rewriter.create<affine::AffineDelinearizeIndexOp>(loc, index, shape)
+        .getMultiIndex();
+  };
+
+  // sync at the end of the loop across threads
+  rewriter.create<gpu::BarrierOp>(loc);
+
+  // For loop body:
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(forOp.getBody());
+    auto inductionVar = forOp.getInductionVar();
+    auto linearizedGatherIndices =
+        getGlobalGatherIndex(subgroupId, laneId, inductionVar);
+    auto delinearizedGlobalIndices =
+        delinearizeIndex(linearizedGatherIndices, sourceType.getShape());
+    auto linearizedBaseIndices =
+        getSubgroupStoreBaseIndex(subgroupId, inductionVar);
+    auto delinearizedLocalIndices =
+        delinearizeIndex(linearizedBaseIndices, localType.getShape());
+    rewriter.create<IREE::GPU::GlobalLoadDMAOp>(
+        loc, copy.getOperand(0), delinearizedGlobalIndices,
+        copy.getOutputs()[0], delinearizedLocalIndices);
+  }
+
+  copy->erase();
+  return true;
+}
+
+} // namespace
+
+static bool checkEligibilityForGlobalLoadDMA(linalg::CopyOp copy) {
+  // source must be global address and target must be workgroup address.
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  auto targetType = cast<MemRefType>(copy.getOutputs().front().getType());
+  if (!hasGlobalMemoryAddressSpace(sourceType)) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has source memory address space other than global.\n");
+    return false;
+  }
+  if (targetType.getMemorySpace() !=
+      gpu::AddressSpaceAttr::get(copy->getContext(),
+                                 gpu::GPUDialect::getWorkgroupAddressSpace())) {
+    LDBG("-- Op: "
+         << *copy
+         << "\n-- has target memory address space other than workgroup.\n");
+    return false;
+  }
+  // TODO: check that the copy's target memref is not a subview.
+  return true;
+}
+
+namespace {
+struct GPULowerToGlobalLoadsPass final
+    : impl::GPULowerToGlobalLoadsPassBase<GPULowerToGlobalLoadsPass> {
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto funcOp = getOperation();
+
+    // we will do this before/at the time of tiling, we need:
+    // 1. tiling configurations, to compute how many to generate.
+
+    SmallVector<linalg::CopyOp> copies;
```

Comment:
The one-shot pattern rewriter might be a good idea here, letting you move all these conditions into a pattern.

Reply:
done.